src/api/StepsExchanger.ts

```diff
@@ -0,0 +1,133 @@
+/*!
+ * The contents of this file are subject to the Common Public Attribution License Version 1.0 (the “License”); you may not use this file except in compliance with the License. You may obtain a copy of the License at https://mpapp-public.gitlab.io/manuscripts-frontend/LICENSE. The License is based on the Mozilla Public License Version 1.1 but Sections 14 and 15 have been added to cover use of software over a computer network and provide for limited attribution for the Original Developer. In addition, Exhibit A has been modified to be consistent with Exhibit B.
+ *
+ * Software distributed under the License is distributed on an “AS IS” basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for the specific language governing rights and limitations under the License.
+ *
+ * The Original Code is manuscripts-frontend.
+ *
+ * The Original Developer is the Initial Developer. The Initial Developer of the Original Code is Atypon Systems LLC.
+ *
+ * All portions of the code written by Atypon Systems LLC are Copyright (c) 2022 Atypon Systems LLC. All Rights Reserved.
+ */
+import { CollabProvider } from '@manuscripts/body-editor'
+import { schema } from '@manuscripts/transform'
+import { Step } from 'prosemirror-transform'
+
+import Api from './Api'
+import { saveWithDebounce } from './savingUtilities'
+
+const MAX_ATTEMPTS = 20
+const THROTTLING_INTERVAL = 1200
+
+export type Subscription = (value: boolean) => void
+
+export class ObservableBoolean {
+  value = false
+  private subs: Subscription[] = []
+
+  setValue(value: boolean) {
+    this.value = value
+    this.subs.forEach((s) => s(value))
+  }
+
+  onChange(sub: Subscription) {
+    this.subs.push(sub)
+  }
+}
+
+export class StepsExchanger extends CollabProvider {
+  projectID: string
+  manuscriptID: string
+  api: Api
+
+  debounce: ReturnType<typeof saveWithDebounce>
+  flushImmediately?: () => void
+
+  isThrottling: ObservableBoolean
+  attempt = 0
+
+  constructor(
+    projectID: string,
+    manuscriptID: string,
+    currentVersion: number,
+    api: Api
+  ) {
+    super()
```

Comment:
In the original implementation there was a check to prevent duplicated instantiation of this class. I would suggest to keep it as there is sometimes little control of what and how react reloads components (especially with upgrades to newer versions). Additional complications may occur in development when react runs each hook twice and hot reload is active.

Reply:
I see. I wasn't aware of that.